### PR TITLE
Ixia port L1 config settings impl

### PR DESCRIPTION
**Pull request by StepanVovkPLV**
Add possibility to change `IXIA `port L1 settings.

### CODE REVIEW

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 1]
```diff
@@ -603,3 +606,23 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
+
+        if not IxnetworkIxiaClientImpl.ixnet:
+            return 0, "Ixia not connected"
+        if command == "update_l1_config":
+            vports = IxnetworkIxiaClientImpl.ixnet.Vport.find()
+            speed = kwarg['params'][0]["speed"]
+            autoneg = kwarg['params'][0].get("autoneg", True)
+            for vport in vports:
+                device.applog.info(f"Changing speed from {vport.ActualSpeed} to {speed} on tgen_port {vport.Name}")
+                # # Currently we have only NovusTenGigLan card in IXIA
```
Comment (StepanVovkPLV) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Will need to add support for any other card

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/ethtool/linux/linux_ethtool_impl.py [hunk 2]
```diff
@@ -100,7 +100,7 @@ def parse_show(self, command, output, *argv, **kwarg):
             line = RE_SPACES.sub(" ", line).strip().split(":")
             key = line[0].replace(" ", "_").lower()
             val = " ".join(line[1:])
-            ethtool_info[key] = val
+            ethtool_info[key.strip()] = val.strip()
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/ethtool/linux/linux_ethtool_impl.py]: this change does not belong to this pr

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/onlp/linux/linux_onlp_system_info_impl.py [hunk 3]
```diff
@@ -64,5 +64,5 @@ def parse_show(self, command, output, *argv, **kwarg):
             line = RE_SPACES.sub(" ", line).strip().split(":")
             key = line[0].replace(" ", "_").lower()
             val = " ".join(line[1:])
-            onl_sys_info[key] = val
+            onl_sys_info[key.strip()] = val.strip()
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/onlp/linux/linux_onlp_system_info_impl.py]: ditto

**Review by SerhiyBoikoPLV — CHANGES_REQUESTED** [review]
Please squash all of your commits together

**Review by SerhiyBoikoPLV — CHANGES_REQUESTED** [review]
please rebase your changes on top of master

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml [hunk 4]
```diff
@@ -33,7 +33,8 @@
           clear_protocol_stats - [protocols]
           send_ping - [port, dst_ip, src_ip]
           send_arp - [port, src_ip]
-    apis: ['connect', 'disconnect', 'load_config', 'save_config', 'set_traffic', 'start_traffic', 'stop_traffic', 'get_stats', 'clear_stats', 'start_protocols', 'stop_protocols', 'set_protocol', 'get_protocol_stats', 'clear_protocol_stats', 'send_arp', 'send_ping']
+          update_l1_config - ['speed', 'autoneg']
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: Please add "duplex" as well. I know that the current card doesn't support this. But we will need it in our testing.

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml [hunk 5]
```diff
@@ -33,7 +33,8 @@
...
+    apis: ['connect', 'disconnect', 'load_config', 'save_config', 'set_traffic', 'start_traffic', 'stop_traffic', 'get_stats', 'clear_stats', 'start_protocols', 'stop_protocols', 'set_protocol', 'get_protocol_stats', 'clear_protocol_stats', 'send_arp', 'send_ping', 'update_l1_config']
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: Maybe there is a reason to split this long line to few smaller lines?

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 1 above)
Comment (anton7811) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: I suggest adding some check that this is NovusGigLan and raise an error is not.

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 6]
```diff
@@ -603,3 +606,23 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
+                      "2500": "speed2.5g",
+                      "5000": "speed5g"}
+
+        if not IxnetworkIxiaClientImpl.ixnet:
+            return 0, "Ixia not connected"
+        if command == "update_l1_config":
+            vports = IxnetworkIxiaClientImpl.ixnet.Vport.find()
+            speed = kwarg['params'][0]["speed"]
+            autoneg = kwarg['params'][0].get("autoneg", True)
+            for vport in vports:
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Can we change the speed mode only on a single port?

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 6 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: not at the moment with this implementation, but it is possible, yes

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml (hunk 4 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: Adding `duplex` depends on the card that supports setting `duplex`. 
At the moment it kind of hard  to do without knowing how many IXIA card will be supported, as all of the have different implemention for the setting of duplex mode(for one card `AutoNegotiation` param is autoneg and `bool` value  for others `AutoNegotiation` param  accepts `string` and is for duplex)

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 7]
```diff
@@ -663,3 +666,30 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
+            return 0, "Ixia not connected"
+        if command == "update_l1_config":
+            vports = IxnetworkIxiaClientImpl.ixnet.Vport.find()
+            ports = kwarg['params'][0].get("ixia_ports", None)
+            speed = kwarg['params'][0].get("speed", None)
+            speed_to_ixia = None if speed is None else speeds_map[str(speed)]
+            autoneg = kwarg['params'][0].get("autoneg", True)
+            names = [vport.Name for vport in vports]
+            for port in ports:
+                if port in names:
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]:
```suggestion
                if port not in names:
                    continue
```

Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: no difference as for me from `if port in names`

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 8]
```diff
@@ -663,3 +666,30 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
+                      "2500": "speed2.5g",
+                      "5000": "speed5g"}
+
+        if not IxnetworkIxiaClientImpl.ixnet:
+            return 0, "Ixia not connected"
+        if command == "update_l1_config":
+            vports = IxnetworkIxiaClientImpl.ixnet.Vport.find()
+            ports = kwarg['params'][0].get("ixia_ports", None)
+            speed = kwarg['params'][0].get("speed", None)
+            speed_to_ixia = None if speed is None else speeds_map[str(speed)]
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: `speed` and `ports` should be mandatory.
1. If the speed that the user passed did not match any speed in the `speeds_map` dict you should return an error.
2. At [L686 ](https://github.com/PLVision/dent-testing/pull/62/files#diff-14cdd5073bf03ba9e02b2ff30bd718aebe2db2d830a399d4093470bf4c8ceeccR686) if `ports` == `None` you will have an exception.

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 7 above)
Comment (SerhiyBoikoPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: There is a difference. You will have less indentation:
```python
for port in ports:
    if port in names:
        required_ixia_port = vports[names.index(port)]
        card = required_ixia_port.L1Config.NovusTenGigLan
        card.update(AutoNegotiate=autoneg, Speed=speed_to_ixia)
```
vs
```python
for port in ports:
    if port not in names:
        continue
    required_ixia_port = vports[names.index(port)]
    card = required_ixia_port.L1Config.NovusTenGigLan
    card.update(AutoNegotiate=autoneg, Speed=speed_to_ixia)
```

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 8 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: 1. `speed ` can be `None` as for IXIA `update()` function it is  it is acceptable param value.(in case we want to change `autoneg `only)
2. Agree on ports thing

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py [hunk 9]
```diff
@@ -743,3 +743,21 @@ async def tgen_utils_send_arp(device, config):
...
+        autoneg (bool): Enable/disable autoneg on the port
+        ixia_ports (list): Ixia ports
+    """
+    device.applog.info("Changing port speed and autoneg")
+    out = await TrafficGen.update_l1_config(input_data=[{device.host_name: [
+        {"speed": speed,
+         "autoneg": autoneg,
+         "ixia_ports": ixia_ports}]}])
+    device.applog.info(out)
+    assert out[0][device.host_name]["result"] == "", f"Failed updating L1 config{out}"
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]:
```suggestion
    assert out[0][device.host_name]["rc"] == 0, f"Failed updating L1 config{out}"
```

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 8 above)
Comment (SerhiyBoikoPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: you didn't resolve the issue with `ports` == `None`

**Review by SerhiyBoikoPLV — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py (hunk 9 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]: we always return `0` and empty string  when executing ixia related code, even on fail. Better to check on `result` here

Comment (SerhiyBoikoPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]: Then don't return 0 on error. Return some error code or -1. Checking `result` here is inconsistent. In other tgen functions you always check for `rc`, so why should you check for `result` here?

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml (hunk 4 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: Added

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py [hunk 10]
```diff
@@ -743,3 +743,23 @@ async def tgen_utils_send_arp(device, config):
     if out[0][device.host_name]["rc"] != 0:
         device.applog.warning(f"Some arps did not reach their destination\n{out}")
     return out[0][device.host_name]["result"]
+
+
+async def tgen_utils_update_l1_config(device, ixia_ports, speed=None, autoneg=True, duplex="Full"):
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]: Please rename `ixia_ports` to `ports`. IMO it is obvious that these are TGen ports. And adding "ixia_" prefix is not acceptable since this is an abstraction, and we can have non-Ixia backend.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py [hunk 11]
```diff
@@ -743,3 +743,23 @@ async def tgen_utils_send_arp(device, config):
...
+    """
+    Update ports L1 config
+    Args:
+        device (DeviceType): Ixia device
+        speed (int| None): Speed to be set on the port
+        autoneg (bool): Enable/disable autoneg on the port
+        ixia_ports (list): Ixia ports
+        duplex (str): Duplex on IXIA port
+    """
+    device.applog.info("Changing port speed and autoneg")
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]: And "duplex"
Maybe better to write: "Change L1 port configuration (autoneg/speed/duplex)"

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml [hunk 12]
```diff
@@ -69,3 +68,19 @@
...
         Peer IP Address
+    - name: speed
+      type: string
+      desc: |
+        Port speed setting
+    - name: autoneg
+      type: bool
+      desc: |
+        Port auto negotiation setting
+    - name: ixia_ports
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: `ixia_ports` -> `ports`

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 13]
```diff
@@ -663,3 +666,42 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
                       "src_ip": msg["src_ip"]} for msg in res]
+
+    def run_update_l1_config(self, device, command, *argv, **kwarg):
+        if not IxnetworkIxiaClientImpl.ixnet:
+            return 0, "Ixia not connected"
+        if command == "update_l1_config":
+            vports = IxnetworkIxiaClientImpl.ixnet.Vport.find()
+            ports = kwarg['params'][0].get("ixia_ports", [])
+            if len(ports) < 1:
+                raise ValueError(f" IXIA ports not provided")
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Please remove the formatting prefix "f" from the string.

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 14]
```diff
@@ -663,3 +666,42 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
+            # fd -> full duplex ; hd -> half duplex
+            duplex = "fd" if kwarg['params'][0].get("duplex", "Full").capitalize() == "Full" else "hd"
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py:673]: Half duplex is default by the standard, but only for 10/100. Maybe it's better to follow the same approach here as well. For 1G/10G without providing duplex explicitly should be "Full".

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 15]
```diff
@@ -663,3 +666,42 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
+                    card = required_ixia_port.L1Config.NovusTenGigLan or required_ixia_port.L1Config.Ethernet
+                    device.applog.info(f"Changing speed from {required_ixia_port.ActualSpeed} to {speed} \
+                                         on tgen_port {required_ixia_port.Name}")
+                    device.applog.info(f"Changing autoneg to {autoneg} on tgen_port {required_ixia_port.Name}")
+                    card.update(AutoNegotiate=autoneg, Speed=ixia_speed)
+            return 0, ""
+
+    @classmethod
+    def __convert_to_ixia_speed(self, speed, duplex):
+        if speed <= 100:
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: "<=" -> `speed == 10 or speed == 100`
So you'll have complete validation of the speed value. Otherwise "99" is also less then 100 :)

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py (hunk 10 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]: Named it first time `ports` but in `traffic.yaml` and `ixia.yaml` wee already have param with the name `ports` and it would not allow to generate the code with non-unique param.  Therefore named it `ixia_ports`. Suggest naming it `tgen_ports`

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py (hunk 11 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]: Indeed looks better

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml (hunk 12 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: See [comment](https://github.com/PLVision/dent-testing/pull/62#discussion_r1139853467)

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 14 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py:673]: IXIA speeds are in three different formats `speed10fd, speed100hd, speed10g ` (endings on **hd, fd , g**)To convert user provided speed to ixia supported speed  use  `__convert_to_ixia_speed(self, speed, duplex)` function

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 15 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Mistake, thank you

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 16]
```diff
@@ -551,3 +552,23 @@ async def tb_device_tcpdump(device, interface, options, count_only=False, timeou
 
     else:
         return out
+
+
+async def tb_get_qualified_ports(device, ports, speed, duplex):
+
+    # Check dut media mode name. If it supports the speed parameter
+    if device.media_mode == "mixed":
+        raise Exception("Mixed mode is not supported")
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Do not use broad Exception, use AttributeError for example

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 17]
```diff
@@ -551,3 +552,23 @@ async def tb_device_tcpdump(device, interface, options, count_only=False, timeou
...
+    if device.media_mode == "copper" and speed == 10000:
+        pytest.skip(reason=f"Can not run test in device {device.host_name}  with the speed: {speed}")
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: We should not skip the test here, better to raise another exception and catch it in the test

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 18]
```diff
@@ -551,3 +552,23 @@ async def tb_device_tcpdump(device, interface, options, count_only=False, timeou
...
+
+    # Check that there are at least minimum 2 ports with the provided speed parameter
+    speed_ports = {}
+    for port in ports:
+        out = await Ethtool.show(input_data=[{device.host_name: [{"devname": port}]}],  parse_output=True)
+        supported_speeds = '{}baseT/{}' if "TP" in out[0][device.host_name]['parsed_output']["supported_ports"] else '{}baseSR/{}'
+        if supported_speeds.format(speed, duplex.capitalize()) in out[0][device.host_name]['result']:
+            speed_ports[port] = {"speed": speed,
+                                 "duplex": duplex}
+    assert len(speed_ports) >= 2, f"Need 2 ports with the same speed of {speed} and duplex {duplex}"
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Add a parameter like "required_ports=None" and if it is set, perform an assertion here

File: DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml [hunk 19]
```diff
@@ -69,3 +68,19 @@
...
+      desc: |
+        Port speed setting
+    - name: autoneg
+      type: bool
+      desc: |
+        Port auto negotiation setting
+    - name: tgen_ports
+      type: list
+      desc: |
+        Ports to change the speed
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbedLib/gen/model/dent/traffic/traffic.yaml]: Not only speed

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 7 above)
Comment (yuriilisovskyi) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Still not resolved

**Review by yuriilisovskyi — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py (hunk 17 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: I asked about this on a call. Anton suggested to skip the test

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py (hunk 7 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Right here(line 676)
` if len(ports) < 1:
                raise ValueError(" IXIA ports not provided")`

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py [hunk 20]
```diff
@@ -663,3 +666,42 @@ def run_send_arp(self, device, command, *argv, **kwarg):
...
                       "src_ip": msg["src_ip"]} for msg in res]
+
+    def run_update_l1_config(self, device, command, *argv, **kwarg):
+        if not IxnetworkIxiaClientImpl.ixnet:
+            return 0, "Ixia not connected"
+        if command == "update_l1_config":
+            vports = IxnetworkIxiaClientImpl.ixnet.Vport.find()
+            ports = kwarg['params'][0].get("tgen_ports", [])
+            if len(ports) < 1:
+                raise ValueError(" IXIA ports not provided")
```
Comment (yuriilisovskyi) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Remove space before IXIA

Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py]: Done

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 21]
```diff
@@ -551,3 +552,36 @@ async def tb_device_tcpdump(device, interface, options, count_only=False, timeou
...
+        required_ports (int): Required number of ports
+
+    Returns: Dictionary of qualified ports
+    """
+
+    # Check dut media mode name. If it supports the speed parameter
+    if device.media_mode == "mixed":
+        raise ValueError("Mixed mode is not supported")
+    if device.media_mode == "copper" and speed == 10000:
+        pytest.skip(reason=f"Can not run test in device {device.host_name} with the speed: {speed}")
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: I don't think that you should skip the test inside of a utils func. This should be done inside of the test itself. If this function is specific to your tests then move it somewhere else so that others won't accidentally skip their tests by calling this func.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py [hunk 22]
```diff
@@ -743,3 +743,23 @@ async def tgen_utils_send_arp(device, config):
...
+
+
+async def tgen_utils_update_l1_config(device, tgen_ports, speed=None, autoneg=True, duplex="Full"):
+    """
+    Update ports L1 config
+    Args:
+        device (DeviceType): Ixia device
+        speed (int| None): Speed to be set on the port
+        autoneg (bool): Enable/disable autoneg on the port
+        tgen_ports (list): Ixia ports
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tgen_utils.py]:
```suggestion
        tgen_ports (list): tgen ports
```

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py (hunk 21 above)
Comment (StepanVovkPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: it will be skipped if if `device.media_mode == "copper" and speed == 10000` because this is not possible technically. User explicitly has to pass speed parameter to function

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 23]
```diff
@@ -551,3 +552,36 @@ async def tb_device_tcpdump(device, interface, options, count_only=False, timeou
...
+    # Check dut media mode name. If it supports the speed parameter
+    if device.media_mode == "mixed":
+        raise ValueError("Mixed mode is not supported")
+    if device.media_mode == "copper" and speed == 10000:
+        raise ValueError(f"Can not run test in device {device.host_name} with the speed: {speed}")
+
+    # Check that there are at least minimum 2 ports with the provided speed parameter
+    speed_ports = {}
+    for port in ports:
+        out = await Ethtool.show(input_data=[{device.host_name: [{"devname": port}]}], parse_output=True)
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py:577]: what if ethtool fails? please check for `rc`

**Review by SerhiyBoikoPLV — APPROVED** [review]
